// [START declaration]
syntax = "proto3";
package sapient_msg.bsi_flex_335_v1_0;

import "sapient_msg/proto_options.proto";
// [END declaration]

// [START java_declaration]
option java_multiple_files = true;
option java_package = "uk.gov.dstl.sapientmsg";
option java_outer_classname = "RangeBearingProto";
// [END java_declaration]

// [START messages]
message RangeBearing {
    optional double elevation = 1 [(field_options) = {xml_name: "Ele"}]; // Elevation in relation to node's horizon
    optional double azimuth = 2 [(field_options) = {xml_name: "Az"}]; // Azimuth in relation to the node's north
    optional double range = 3 [(field_options) = {xml_name: "R"}]; // Range from the node's location
    optional double elevation_error = 4 [(field_options) = {xml_name: "eEle"}]; // Elevation error
    optional double azimuth_error = 5 [(field_options) = {xml_name: "eAz"}]; // Azimuth error
    optional double range_error = 6 [(field_options) = {xml_name: "eR"}]; // Range Error
    optional RangeBearingCoordinateSystem coordinate_system = 7 [(field_options) = {is_mandatory: true, xml_ignore: true}]; // Units used by the range bearing
    optional RangeBearingDatum datum = 8 [(field_options) = {is_mandatory: true, xml_ignore: true}]; // Datum or 'north' used by the range bearing
}

message RangeBearingCone {
    optional double elevation = 1 [(field_options) = {xml_name: "Ele"}]; // Elevation in relation to node's horizon
    optional double azimuth = 2 [(field_options) = {xml_name: "Az"}]; // Azimuth in relation to the node's north
    optional double range = 3 [(field_options) = {xml_name: "R"}]; // Range from the node's location
    optional double horizontal_extent = 4 [(field_options) = {xml_name: "hExtent"}]; // Horizontal extent angle in degrees
    optional double vertical_extent = 5 [(field_options) = {xml_name: "vExtent"}]; // Vertical extent angle in degrees
    optional double horizontal_extent_error = 6 [(field_options) = {xml_name: "ehExtent"}]; // Error in the horizontal extent
    optional double vertical_extent_error = 7 [(field_options) = {xml_name: "evExtent"}]; // Error in the vertical extent
    optional double elevation_error = 8 [(field_options) = {xml_name: "eEle"}]; // Elevation error
    optional double azimuth_error = 9 [(field_options) = {xml_name: "eAz"}]; // Azimuth error
    optional double range_error = 10 [(field_options) = {xml_name: "eR"}]; // Range Error
    optional RangeBearingCoordinateSystem coordinate_system = 11 [(field_options) = {is_mandatory: true, xml_ignore: true}]; // Units used by the range bearing
    optional RangeBearingDatum datum = 12 [(field_options) = {is_mandatory: true, xml_ignore: true}]; // Datum or 'north' used by the range bearing
}

enum RangeBearingCoordinateSystem {
    reserved 5, 6; // These values were used in SAPIENT upto version 7, but are no longer used due to use of non-SI units
      RANGE_BEARING_COORDINATE_SYSTEM_UNSPECIFIED = 0; // Co-ordinate system/units not defined
      RANGE_BEARING_COORDINATE_SYSTEM_DEGREES_M = 1 [(enum_options) = {enum_name: "decimal degrees-meters"}]; // Values in decimal-degrees and meters
      RANGE_BEARING_COORDINATE_SYSTEM_RADIANS_M = 2 [(enum_options) = {enum_name: "radians-meters"}]; // Values in radians and meters
      RANGE_BEARING_COORDINATE_SYSTEM_DEGREES_KM = 3 [(enum_options) = {enum_name: "decimal degrees-kilometers"}]; // Values in decimal-degrees and kilometers
      RANGE_BEARING_COORDINATE_SYSTEM_RADIANS_KM = 4 [(enum_options) = {enum_name: "radians-kilometers"}]; // Values in radians and kilometers
//    RANGE_BEARING_COORDINATE_SYSTEM_DEGREES_F = 5; // Values in decimal-degrees and feet - DEPRECATED
//    RANGE_BEARING_COORDINATE_SYSTEM_RADIANS_F = 6; // Values in radians and feet - DEPRECATED
}

enum RangeBearingDatum {
    RANGE_BEARING_DATUM_UNSPECIFIED = 0; // Datum not defined
    RANGE_BEARING_DATUM_TRUE = 1; // True North
    RANGE_BEARING_DATUM_MAGNETIC = 2; // Magnetic North
    RANGE_BEARING_DATUM_GRID = 3; // Grid North
    RANGE_BEARING_DATUM_PLATFORM = 4; // 'North' is the heading of the platform carrying the node
}
// [END messages]
